strawberryfields/program.py

```diff
@@ -436,27 +447,44 @@ def _linked_copy(self):
             p.source = self.source
         return p
 
-    def compile(self, target, **kwargs):
-        """Compile the program targeting the given circuit specification.
+    def compile(self, device_or_compiler, force_compiler=None, **kwargs):
+        """Compile the program given a Strawberry Fields photonic compiler, or
+        hardware device specification.
+
+        The compilation process can involve up to three stages:
+
+        1. **Validation:** Validates properties of the program, including number of modes and
+           allowed operations, making sure all the :doc:`/introduction/ops` used are accepted by the
+           compiler.
 
-        Validates the program against the given target, making sure all the
-        :doc:`/introduction/ops` used are accepted by the target specification.
+        2. **Decomposition:** Once the program has been validated, decomposition are performed,
+           transforming certain gates into sequences of simpler gates.
 
-        Additionally, depending on the target, the compilation may modify the quantum circuit
-        into an equivalent circuit, e.g., by decomposing certain gates into sequences
-        of simpler gates, or optimizing the gate ordering using commutation rules.
+        3. **General compilation:** Finally, the compiler might specify bespoke compilation logic
+           for transforming the  quantum circuit into an equivalent circuit which can be executed
+           by the target device.
 
         **Example:**
 
         The ``gbs`` compile target will
         compile a circuit consisting of Gaussian operations and Fock measurements
         into canonical Gaussian boson sampling form.
 
-        >>> prog2 = prog.compile('gbs')
+        >>> prog2 = prog.compile("gbs")
+
+        For a hardware device a :class:`~.DeviceSpec` object, and optionally a specified compile strategy,
+        must be supplied. If no compile strategy is supplied the default compiler from the device
+        specification is used.
+
+        >>> eng = sf.RemoteEngine("X8")
+        >>> device = eng.device_spec
+        >>> prog2 = prog.compile(device, "Xcov")
 
         Args:
-            target (str, ~strawberryfields.circuitspecs.CircuitSpecs): short name of the target
-                circuit specification, or the specification object itself
+            device_or_compiler (str, ~strawberryfields.compilers.Compiler, ~strawberryfields.api.DeviceSpec):
+                compiler name or device specification object to use for program compilation
+            force_compiler (str, ~strawberryfields.compilers.Compiler): Optionally provide a compile strategy.
+                This overrides the compile strategy specified by a hardware :class:`~.DevicSpec`.
```

Comment:
Having the ability to pass a `~strawberryfields.compilers.Compiler` to both `device_or_compiler` and `force_compiler` seems a little bit odd.

Would having a joint `device_options` dictionary be worth considering here? One that can be passed as a positional argument. It could have two keys: `"device"` and `"compile"`. At least one would have to be in the dictionary, and if both are, then the `compile` would overwrite the compilation strategy (the current `force_compiler` option).

Reply:
I agree that this is a bit confusing and odd. Especially since a user could (but shouldn't) set `device_or_compiler` to a compiler _and_ add another `force_compiler` as well, which would simply just not do anything.

Reply:
I agree that it is confusing, and struggled a bit here.

Perhaps one solution is to use the Python `*` syntax, so that positional arguments are not allowed. Then we could do something like

```python
def compile(self, *, device=None, compiler=None):
```

A user could then optionally pass one or the other. This avoids the weird issue with mutually exclusive positional arguments!

Reply:
What would happen if both `device` and `compiler` are left as `None`?

Reply:
I think that the suggestion from @josh146 might actually be quite nice. Having both `device` and `compiler` as `None` could perhaps simply just use a default compiler (either `Xcov` or `Xunitary`)?